scrapy/utils/misc.py

```diff
@@ -116,4 +116,28 @@ def md5sum(file):
 def rel_has_nofollow(rel):
     """Return True if link rel attribute has nofollow type"""
     return True if rel is not None and 'nofollow' in rel.split() else False
-    
+
+def construct_from_settings_or_crawler(objcls, settings, crawler, *args,
```

Comment:
[Here](https://github.com/scrapy/scrapy/blob/master/scrapy/crawler.py#L314) is a use case where we cannot deprecate `from_settings()`, but it should be possible for middlewares/pipelines/extensions. Maybe we could call it `construct_from_crawler()` and add a deprecation note that triggers when `from_settings()` exists, but `from_crawler()` does not?